### ISSUE TITLE
Can we import module conditionally?

### ISSUE THREAD

**Issue by bakura10** [guybedford/es-module-shims#371]
Hi :),

This is not a bug but rather a question on how to embed the script. We are using this only for import map support. For now, we have followed the embed instructions to use an async script. As all browsers now support it natively, we would prefer to load it conditionally (we're not saving a lot, I know, but we are selling themes and a lot of our users are running their stores on PageSpeed, and those non-technical merchants often ask us how to "optimize their store by removing scripts" ; we therefore try to load as little JS as possible to avoid support debt).

I tried the following approach which works on all the iOS versions I have tested:

```
<script>
      if (!HTMLScriptElement.supports('importmap')) {
        import('URL-TO-POLYFILL');
      }
    </script>
```

However, I am not 100% sure about possible consequences or potential timing issue. Considering this polyfill architecture, would this be as safe as embedding it using the async way?

Thanks.

**Comment by guybedford** [comment]
Agreed that now that we have full browser support conditional loading would be worth fleshing out more clearly.

This is something that would need to be designed as a core feature of the project because we have feature detection scripts which would still need to run. So effectively we would change the structure of es-module-shims to first include a simple feature detection script, that in turn would dynamically import the main polyfill only when needed.

There is still a tradeoff on that for a little while while browsers that don't support import maps would be penalized by that extra round trip. But as import maps gain in widespread usage this seems like a must-have to me.

**Comment by bakura10** [comment]
For those trying the approach I have outlined (dynamically importing the module) this does not work all the time. I have timing issues, so I will wait for a built-in approach :D.

**Comment by bakura10** [comment]
For anyone interested in doing that, to make it work you actually need to append the element to the dom (and not try to dynamically importing it). This approach works to conditionally load the script:

```
<script>
      if (!(HTMLScriptElement.supports && HTMLScriptElement.supports('importmap'))) {
        const importMapPolyfill = document.createElement('script');
        importMapPolyfill.async = true;
        importMapPolyfill.src = "URL-TO-POLYFILL";

        document.head.appendChild(importMapPolyfill);
      }
    </script>
```

**Comment by guybedford** [comment]
It seems like conditionally loading is working pretty well as Wordpress landed in https://github.com/WordPress/gutenberg/pull/57256/files using a `HTMLScriptElement.supports('importmap')` and `document.write` approach.

Perhaps the solution here is then just simply to document this as a first-class feature and continue supporting it as the primary approach going forward.

**Comment by luisherranz** [comment]
> Perhaps the solution here is then just simply to document this as a first-class feature and continue supporting it as the primary approach going forward

That sounds good to me!

To be honest, I don't know why `document.head.appendChild` didn't work in our case. Maybe I did something wrong. I want to try again.

Or does it make sense that `document.write` works, but `document.head.appendChild` doesn't?

**Comment by guybedford** [comment]
I'm going to need some time to properly dig into the list of DOM attachments and implications for dynamic attachment. The current assumptions for the shim are very much that it's there first for performance reasons and is part of the loading lifecycle of the page (refiring dom events as necessary for the new loaded scripts). I suppose at least the `document.write` applies only in the legacy cases now so shouldn't apply for modern browsers at least, but agreed it would be preferable to figure out.

I'm going to track this issue and get back with further details when I find some time. In the mean time contributions are of course welcome too.

**Comment by luisherranz** [comment]
> I suppose at least the `document.write` applies only in the legacy cases now so shouldn't apply for modern browsers at least

Yes. We don't mind releasing it in WordPress with `document.write` because it will only affect a small and shrinking percentage of users. So it's not critical at all. But I wanted to understand why it needs to use that method before committing to it.

> The current assumptions for the shim are very much that it's there first for performance reasons and is part of the loading lifecycle of the page

There's one part that still doesn't make sense to me. Your recommendation is:

`<script async src="es-module-shims.js"></script>`

Async scripts can load at any time. [They are not even guaranteed to load before the `DOMContentLoaded` event](https://developer.mozilla.org/en-US/docs/Web/API/Document/DOMContentLoaded_event#:~:text=The%20DOMContentLoaded%20event%20fires%20when%20the%20HTML%20document%20has%20been%20completely%20parsed%2C%20and%20all%20deferred%20scripts%20(%3Cscript%20defer%20src%3D%22%E2%80%A6%22%3E%20and%20%3Cscript%20type%3D%22module%22%3E)%20have%20downloaded%20and%20executed.%20It%20doesn%27t%20wait%20for%20other%20things%20like%20images%2C%20subframes%2C%20and%20async%20scripts%20to%20finish%20loading.). 

So, at least theoretically, there should be no difference between an async script and the `document.head.appendChild` method.
